### DISCUSSION TITLE
Integrating LibreChat's Frontend with Custom Python Backend Using OpenAI's API

### DISCUSSION THREAD

**Discussion by gjgeorgiadis** [danny-avila/LibreChat#2839] [Q&A]
Hello everyone,

I have a chatbot programmed in Python that utilizes OpenAI's API. Currently, the frontend is a simple HTML page, which lacks both aesthetic appeal and functionality such as maintaining query history and a search function.

I am exploring the possibility of using LibreChat's frontend with my existing backend. My current backend involves an algorithm that sorts documents to find the ones closest to the user's query, edits the query to add any missing definitions/clarifications, and then feeds it to GPT-4.

I attempted to replicate my chatbot in LibreChat using an OpenAI assistant, but the results were poor. I suspect this is because my custom algorithm sorts documents to find the ones closest to the user's query and edits the query to add any missing definitions/clarifications before feeding it to GPT4.

**Questions:**

1. Is it feasible to integrate LibreChat's frontend with my custom Python backend that runs off OpenAI's API?
2. If so, what would be the best approach to achieve this integration?
3. Are there specific challenges or limitations I should be aware of when attempting this integration?

Any advice or pointers would be greatly appreciated!
Thank you.

**Comment by danny-avila** [comment]
Make your Python API an OpenAI compatible API, and add it as a Custom Endpoint:
https://www.librechat.ai/docs/configuration/librechat_yaml/object_structure/custom_endpoint

That would be the best/easiest/most seamless move.

You can test compatibility with the openai-node library, which we use: https://github.com/openai/openai-node

**Reply by gjgeorgiadis** [reply]
Thank you very much Danny!

**Reply by danny-avila** [reply]
You're welcome!

**Reply by usnavy13** [reply]
I just came across this, I can confirm Danny's suggested solution works. i have implemented it with streaming support. I am still working on asyc support. Will share a code sample if i remember on Monday.

**Reply by wewesmog** [reply]
Hello, I able to create a working OpenAI compatible endpoint. However, I dont know the steps to follow to have it working on librechat. Kindly guide: See my code below for testing the API:
import requests

def test_chat():
    url = "http://212.XX.XX.75:8001/v1/chat/completions"
    
    headers = {
        "Content-Type": "application/json"
    }
    
    payload = {
        "model": "custom-model",
        "messages": [
            {
                "role": "user",
                "content": "What is Okolea loan?"
            }
        ]
    }

    try:
        print(f"Sending request to: {url}")
        print(f"Payload: {payload}")
        
        response = requests.post(url, headers=headers, json=payload)
        print(f"Status Code: {response.status_code}")
        print("Response Headers:", dict(response.headers))
        print("Response Body:", response.text)
        
        if response.status_code == 200:
            print("Parsed JSON Response:", response.json())
            
    except requests.exceptions.RequestException as e:
        print(f"Request Error: {e}")
    except Exception as e:
        print(f"Unexpected Error: {e}")

if __name__ == "__main__":
    test_chat()

**Reply by danny-avila** [reply]
You can test compatibility with the openai-node library, which we use: https://github.com/openai/openai-node
